### PR TITLE
HSEARCH-4542 .within() and .withinAny() for the range predicate

### PR DESCRIPTION
**Pull request by marko-bekhta**
https://hibernate.atlassian.net/browse/HSEARCH-4542

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/engine/search/predicate/dsl/RangePredicateMatchingStep.java [hunk 1]
```diff
@@ -225,7 +227,50 @@ default N range(Range<?> range) {
...
+	 * See {@link ValueConvert} for more information.
+	 * @param convert Controls how the range bounds should be converted
+	 * before Hibernate Search attempts to interpret them as a field value.
+	 * See {@link ValueConvert} for more information.
+	 * @return The next step.
+	 */
+	N within(Range<?> range, ValueConvert convert);
+
+	/**
+	 * Require at least one of the targeted fields to be in the given range.
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/search/predicate/dsl/RangePredicateMatchingStep.java:265]:
```suggestion
	 * Require at least one of the targeted fields to be in any of the given ranges.
```

File: engine/src/main/java/org/hibernate/search/engine/search/predicate/dsl/RangePredicateMatchingStep.java [hunk 2]
```diff
@@ -225,7 +227,50 @@ default N range(Range<?> range) {
...
+	 *
+	 * @param ranges The ranges to match.
+	 * The signature of this method defines this parameter as a range with bounds of any type,
+	 * but a specific type is expected depending on the targeted field.
+	 * See {@link ValueConvert#YES} for more information.
+	 * @return The next step.
+	 */
+	N withinAny(Range<?>... ranges);
 
+	RangePredicateMatchingWithinAnyMoreStep<?> withinAny();
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/search/predicate/dsl/RangePredicateMatchingStep.java:275]: Unless you already found similar syntax elsewhere, I would recommend sticking to `withinAny(Range<?>...)` / `withinAny(Collection<? extends Range<?>>)` / `withinAny(Collection<? extends Range<?>>, ValueConvert)`.

See `org.hibernate.search.engine.search.predicate.dsl.TermsPredicateMatchingStep`.

If people want to use a different `ValueConvert` for each range, let them find a solution for themselves. There are limits to how flexible we should be, IMO.

**Review by yrodiere — COMMENTED** [review]
Thanks; a few comments below.

FWIW, I think aligning on the syntax of the `terms` predicate is better than aligning on the syntax of the `range` aggregation.

I even wonder if we should change the syntax of the `range` aggregation... This `.range(....).range(...)` is quite awkward, too. Not in this PR for sure, though :)

File: engine/src/main/java/org/hibernate/search/engine/search/predicate/dsl/RangePredicateMatchingStep.java (hunk 2 above)
Comment (marko-bekhta) [reply engine/src/main/java/org/hibernate/search/engine/search/predicate/dsl/RangePredicateMatchingStep.java:275]: Thanks for taking a look at this one! 😃 👍🏻 ok will do😃

**Review by yrodiere — APPROVED** [review]

**Pull request merged**